godot-core/src/builtin/quaternion.rs

```diff
@@ -130,70 +138,115 @@ impl Quaternion {
-            b_t: real,
-            pre_a_t: real,
-            post_b_t: real,
-        ) -> Self {
-        }
-    */
+        let interpolated = self.as_inner().slerpni(to, weight.as_f64());
+
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let normalized_inputs = !interpolated.is_default()
+            || interpolated.is_default() && self.ensure_normalized(&[&to]);
+        assert!(normalized_inputs, "Slerpni requires normalized quaternions");
+
+        interpolated
+    }
+
+    /// # Panics
+    /// If any quaternions are not normalized.
+    pub fn spherical_cubic_interpolate(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+    ) -> Self {
+        let interpolated =
+            self.as_inner()
+                .spherical_cubic_interpolate(b, pre_a, post_b, weight.as_f64());
+
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let normalized_inputs = !interpolated.is_default()
+            || interpolated.is_default() && self.ensure_normalized(&[&b, &pre_a, &post_b]);
+        assert!(
+            normalized_inputs,
+            "Spherical cubic interpolation requires normalized quaternions"
+        );
+
+        interpolated
+    }
+
+    /// # Panics
+    /// If any quaternions are not normalized.
+    #[allow(clippy::too_many_arguments)]
+    pub fn spherical_cubic_interpolate_in_time(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+        b_t: real,
+        pre_a_t: real,
+        post_b_t: real,
+    ) -> Self {
+        let interpolated = self.as_inner().spherical_cubic_interpolate_in_time(
+            b,
+            pre_a,
+            post_b,
+            weight.as_f64(),
+            b_t.as_f64(),
+            pre_a_t.as_f64(),
+            post_b_t.as_f64(),
+        );
+
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let normalized_inputs = !interpolated.is_default()
+            || interpolated.is_default() && self.ensure_normalized(&[&b, &pre_a, &post_b]);
+        assert!(
+            normalized_inputs,
+            "Spherical cubic interpolation in time requires normalized quaternions"
+        );
+        interpolated
+    }
 
     #[doc(hidden)]
     pub fn as_inner(&self) -> inner::InnerQuaternion {
         inner::InnerQuaternion::from_outer(self)
     }
+
+    #[doc(hidden)]
+    pub fn is_default(&self) -> bool {
+        *self == Self::default()
+    }
+
+    #[doc(hidden)]
+    pub fn ensure_normalized(&self, quats: &[&Quaternion]) -> bool {
+        quats.iter().all(|v| v.is_normalized()) && self.is_normalized()
+    }
```

Comment:
Why are these `pub` and `#[doc(hidden)]`? Could just be private functions, no?

Reply:
Good callout.